### DISCUSSION TITLE
Testing Time-Dependent Withdrawal in Extended FundMe Project: Seeking Testing Guidance

### DISCUSSION THREAD

**Discussion by shahin-bayat** [Cyfrin/foundry-full-course-cu#742] [Q&A] [answered]
Hello Foundry Solidity community,

I've been extending the FundMe project and now have a feature where the owner can withdraw funds after a specific time period. I'm excited about this progress, but I'm facing a challenge in testing this functionality effectively.

The scenario involves ensuring that the withdrawal function only works after a set time and under certain conditions. I've considered using the block.timestamp and require statements to manage timing, but I'm unsure about the best way to test this setup.

If you've encountered similar situations or have insights on testing time-dependent contract functions, could you kindly share your experiences? I'd greatly appreciate any guidance or best practices you can offer.

Thank you for your support and expertise!

```solidity
    function withdrawFunds() public onlyProjectOwnner {
        require(block.timestamp > s_campaignDeadline, "Campaign deadline not passed yet");
        require(s_totalFunding >= i_fundingGoal, "Funding goal not reached");
        (bool callSuccess, ) = payable(i_projectOwner).call{value: s_totalFunding}("");
        require(callSuccess, "Withdraw funds failed");
    }
```

**Comment by shikhar229169** [comment, marked as answer]
We have a cheatcode to go ahead in time and that is used in the later lesson.
```solidity
vm.warp(x)
```
So warp will set the current timestamp to x. So, before withdrawing you can use this cheatcode to get ahead in time.
Refer to the docs: https://book.getfoundry.sh/cheatcodes/warp

**Reply by shahin-bayat** [reply]
Thank you so much.
